### PR TITLE
V0.4 addcommand smart parsing

### LINKED ISSUE #182: Feature Suggestion : allow multiple ways to input dates
`floating task` : 
- no date flags

`deadline` :
- only `by` 

`event` :
- `to` or `by` , must be used in conjunction with `from`,`at`,`on` 
- `from`, `at`, `on` used alone with default +1hour duration (from config) to infer enddatetime

anything else : parse error

### PR DESCRIPTION
**Pull request by kevinLamKB**
Changelog 📜 
- implement smart parsing for add commands.
- `from`,`at`,`on` can be used alone to signify events, currently adding a magic number 1 hour for endDate, will update to config addDurationHour when its ready.
- `to` and `by` must be used together with a start date if user wants to add event with his specified enddate
- update to parsing logic, can now use any combination of flags to denote event as long as start and end is given, parser will only take the last end and start flag. e.g `add someevent from thursday by noon to 1 day by 6000 years ago on wat from today to tomorrow` will use `from today` and `to tomorrow`

- moved one of mattheus's method to ArgumentMultimap to allow reuse for my other parsers, update collated tag and listcommandparser to follow

Fixes #182

### CODE REVIEW

File: src/main/java/seedu/multitasky/logic/parser/ArgumentMultimap.java [hunk 1]
```diff
@@ -60,4 +60,20 @@ public void put(Prefix prefix, String argValue) {
...
+     * Method to extract an ArrayList of specified prefixes that are present in the ArgumentMultimap
+     */
+    public ArrayList<String> getPresentPrefixes(String... prefixes) {
+        ArrayList<String> presentPrefixes = new ArrayList<>();
+        for (String prefix : prefixes) {
+            if (getAllValues(new Prefix(prefix)).size() != 0) {
+                presentPrefixes.add(prefix);
+            } else {
+                // continue
+            }
```
Comment (mattheuslee) [inline src/main/java/seedu/multitasky/logic/parser/ArgumentMultimap.java]: I think there's no need for the `else` block, I originally had it as I was doing a println there while debugging, but now it's not needed anymore.

**Comment by mattheuslee** [comment]
> update to parsing logic, can now use any combination of flags to denote event as long as start and end is given, parser will only take an end and start flag of the last entry. e.g add someevent from thursday by noon to 1 day by 6000 years ago on wat from today to tomorrow will use from today and to tomorrow
 
What an example 😆

**Review by mattheuslee — APPROVED** [review]

**Review by akd13 — APPROVED** [review]

**Review by ChuaPingChan — APPROVED** [review]
@CS2103JUN2017-T2/developers 
Some of the previous tests will need to be modified after this PR is merged it seems.

File: src/main/java/seedu/multitasky/logic/parser/ArgumentMultimap.java (hunk 1 above)
Comment (kevinLamKB) [reply src/main/java/seedu/multitasky/logic/parser/ArgumentMultimap.java]: noted, im removing it

**Pull request merged**